### ISSUE TITLE
Additional, reduced round ChaCha ciphers

### ISSUE THREAD

**Issue by Monn-i** [openssl/openssl#25347]
As detailed in [Aumasson, J-P. (2020) Too Much Crypto. Real World Crypto Symposium.](https://eprint.iacr.org/2019/1492), ChaCha20 cipher may have been "overshot" with its rounds and security margins.

Would it be possible to provide reduced round variants of the ChaCha20? As per paper, the speed-up for ChaCha8 could be approximately 2.5 times compared to the 20-round variant and it's yet to be proven unsecure - so far the most rounds proven to be unsecure is 7 due to the exponentially harder to counter diffusion with the rounds. This speed-up would have considerable benefits especially for low-performance platforms like many embedded systems are.

Side note: This might be something I also could contribute to. By the looks of it this only reduces the algorithm rounds, without touching anything other in the algorithm implementation. Does this seem like a relevant thing to add to OpenSSL library?

**Comment by paulidale** [comment]
The barrier to inclusion of an algorithm in OpenSSL is being a national or international standard.  Does this meet that level?

**Comment by Monn-i** [comment]
I looked into publications on this and I admit it hasn't been consolidated anywhere enough to say it has a standard status. IEEE has some papers approaching these ciphers, most direct being [A Hardware Perspective on the ChaCha Ciphers: Scalable Chacha8/12/20 Implementations Ranging from 476 Slices to Bitrates of 175 Gbit/s](https://ieeexplore.ieee.org/document/9088102).

With this info I understand this doesn't pass the barrier of inclusion by standard consolidation. I wouldn't mind OTC having a say on the subject, but won't press this feature request further personally for the reasons given.

**Comment by paulidale** [comment]
I've flagged this for OTC discussion.

OTC: **should we consider reduced round algorithms for inclusion?**

They could be implemented either as a new algorithm or via a OSSL_PARAM specifying the number of rounds.

**Comment by hlandau** [comment]
**OTC: In the absence of a national or international standard specifying the use of reduced-round ChaCha, we recommend this be implemented as a third-party provider.**

**Comment by Monn-i** [comment]
Thanks for the informative thoughts, 3rd party provider for the ciphers sounds like a good approach even if OpenSSL itself won't incorporate the ciphers yet.

> Thus, it does not really make sense to support it without optimised assembly implementations for the reduced-round variants. This would be a significant addition of code for something which is not yet standardised.

I too agree with this after hearing the reasoning. From my part, I see this request resolved as "not planned", I'll look into implementing this as a 3rd party provider.

**Comment by SammyVimes** [comment]
@hlandau I'd like to clarify a thing regarding the assembly-optimized versions of `ChaCha20_ctr32`. While I’m not an assembly expert, it seems possible to make the number of rounds a parameter for this function. If I’m not mistaken, the round count is stored in a register and then decremented, like in any loop. I understand that loops can be unrolled and this way round number does indeed matter, but it doesn't seem to be the case here.

As a test, I replaced `$10` with `$4` (for ChaCha8), and the resulting ciphertext aligned with other ChaCha8 implementations.

Basically I am just curious about why in your opinion the only possible option is a portable version of chacha generic

**Comment by paulidale** [comment]
Did you do this for **all** of the assembly implementations or just one?
Some look like they've been unrolled.

**Comment by SammyVimes** [comment]
> Did you do this for **all** of the assembly implementations or just one? Some look like they've been unrolled.

For all, but only in linux implementation (since that it is assembly I know to some extent). I see the unrolling, but it looks like it repeats cycle body twice, so I *think* that chacha8 and chacha12 (or any even number of rounds) can be done without copying everything. 

It is not in any way a call to any action, just my genuine curiosity

**Comment by paulidale** [comment]
Linux versions run on lots of different processors.  I'm assuming you only checked x86.

**Comment by Artoria2e5** [comment]
20 is 4 times 5, so it stands to reason that any unrolling won't be that bad to re-size for accomodating the other round counts...
